src/sentry/ratelimits/cardinality.py

```diff
@@ -132,3 +157,251 @@ def use_quotas(
+        unit_keys_to_get: List[str] = []
+        set_keys_to_count: List[str] = []
+
+        for request in requests:
+            if request.quotas:
+                for hash in request.unit_hashes:
+                    unit_keys_to_get.append(self._get_timeseries_key(request, hash))
+
+            for quota in request.quotas:
+                set_keys_to_count.extend(self._get_read_sets_keys(request, quota, timestamp))
+
+        if not unit_keys_to_get and not set_keys_to_count:
+            # If there are no keys to fetch (i.e. there are no quotas to
+            # enforce), we can save the redis call entirely and just grant all
+            # quotas immediately.
+            return timestamp, [
+                GrantedQuota(
+                    request=request, granted_unit_hashes=request.unit_hashes, reached_quotas=[]
+                )
+                for request in requests
+            ]
+
+        with self.client.pipeline(transaction=False) as pipeline:
+            pipeline.mget(unit_keys_to_get)
+            # O(self.cluster_shard_factor * len(requests)), assuming there's
+            # only one per-org quota
+            for key in set_keys_to_count:
+                pipeline.scard(key)
+
+            results = iter(pipeline.execute())
+            unit_keys = dict(zip(unit_keys_to_get, next(results)))
+            set_counts = dict(zip(set_keys_to_count, results))
+
+        grants = []
+        cardinality_sample_factor = self._get_set_cardinality_sample_factor()
+        for request in requests:
+            if not request.quotas:
+                grants.append(
+                    GrantedQuota(
+                        request=request, granted_unit_hashes=request.unit_hashes, reached_quotas=[]
+                    )
+                )
+                continue
+
+            granted_hashes = []
+
+            quotas_by_remaining_limit = defaultdict(list)
+            for quota in request.quotas:
+                remaining_limit = max(
+                    0,
+                    quota.limit
+                    - cardinality_sample_factor
+                    * sum(
+                        set_counts[k] for k in self._get_read_sets_keys(request, quota, timestamp)
+                    ),
+                )
+                quotas_by_remaining_limit[remaining_limit].append(quota)
+
+            smallest_remaining_limit = min(quotas_by_remaining_limit)
+            smallest_remaining_limit_running = smallest_remaining_limit
+            reached_quotas = []
+
+            for hash in request.unit_hashes:
+                if unit_keys[self._get_timeseries_key(request, hash)]:
+                    granted_hashes.append(hash)
+                elif smallest_remaining_limit_running > 0:
+                    granted_hashes.append(hash)
+                    smallest_remaining_limit_running -= 1
+                else:
+                    reached_quotas = quotas_by_remaining_limit[smallest_remaining_limit]
+                    break
```

Comment:
Could you explain what is happening in this code. I feel this is little hard to comprehend. Maybe adding some comments in the code would be helpful.